db/version_set.cc

```diff
@@ -571,14 +629,24 @@ class LevelIterator final : public InternalIterator {
   RangeDelAggregator* range_del_agg_;
   IteratorWrapper file_iter_;  // May be nullptr
   PinnedIteratorsManager* pinned_iters_mgr_;
+  RangeTombstone range_tombstone_;
 };
 
 void LevelIterator::Seek(const Slice& target) {
   size_t new_file_index = FindFile(icomparator_, *flevel_, target);
 
   InitFileIterator(new_file_index);
   if (file_iter_.iter() != nullptr) {
-    file_iter_.Seek(target);
+    // Before we seek the iterator, find the next non-deleted key.
```

Comment:
Placing this awareness of range tombstones in `LevelIterator` works, but I think it might be the wrong place. If this were added to `BlockBasedTableIterator` then we'd get the benefit for L0 files as well.

Reply:
Is there a reason to prefer that approach vs plumbing this range tombstone skipping down into `BlockBasedTableIterator`?

Reply:
One benefit of the LevelIterator approach would be avoiding opening tables that we aren't going to read any data from. Although, for users who set table cache's capacity to infinity (like us), that wouldn't matter much.

I talked to @siying and he proposed another idea for your consideration. We could put the logic for seeking past a range tombstone in the `MergingIterator`. That way we can seek directly to the tombstone end key without processing files that are skipped over one-by-one. It also can allow us to write generic code that handles all levels including L0.

One thing we'd need is an interface in `InternalIterator` for getting max seqnum. `LevelIterator::GetMaxSeqnum` (which would be used for L1+) can return the max of max seqnums of all files in the level. Then we'd only `Seek` to a range tombstone's end key if the tombstone's seqnum is greater. This is a bit less optimal than your previous approach as there'll be slightly fewer cases where `Seek` can be used.

For `BlockBasedTableIterator::GetMaxSeqnum`, which would by `MergingIterator` for its L0 children, it can simply return that table's max seqnum. Logic for seeking should be the same. 

What do you think of this approach?

Reply:
I need to think about this idea more. I see the attraction to putting this logic in `MergingIterator` and that was actually where I was initially considering putting logic to skip tombstones. But I couldn't figure out how to provide information about sstable max sequence numbers. The suggestion to only consider the max sequence number for the entire level is interesting. I wonder if there are scenarios where the max sequence number for one part of a level is vastly different than for another part of the level, defeating the range tombstone skipping optimization.

Reply:
Yeah that's true, there can be compaction hotspots within a level, and files in those hotspots can have much newer seqnums.

I'm not sure about your compaction config, but if you're using level compaction with fanout of 10 like we are, typically ~90% of the data will be in the bottom level. We reset most seqnums to zero in the bottom level when we determine there are no older versions of the key in the DB. Then the optimization, whether implemented in MergingIterator or BlockBasedTableIterator, I think should apply to the ~90% of data in bottom level.

Reply:
There are a couple known cases where we don't reset seqnum to zero during compaction to bottom level:

- when a snapshot is held and the output file contains a key newer than the snapshot
- when we output the final key in the file. IMO this is a bug that we should fix

Reply:
@ajkr Do you have a preference to placing this optimization in `BlockBasedTableIterator` or `MergingIterator`? I'm leaning towards `BlockBasedTableIterator`. It avoids having to add `InternalIterator::GetMaxSeqnum` while providing at least as much benefit as adding the optimization to `MergingIterator`. I haven't started on implementation though, so I might change my mind once I do.

Reply:
I'd be ok with either. In `BlockBasedTableIterator` there's a small downside to people setting `max_open_files != -1` because they'll have to do some I/O on files that'll be entirely skipped over. There's also the case where range deletion is huge and spans many files, so the sequential check over files is less efficient than the binary search we could do in `MergingIterator`. But either way it'll be much better than what we have now so I'd accept whichever way you like :).